### DISCUSSION TITLE
Domino-ui 2.x.x migration tips

### DISCUSSION THREAD

**Discussion by vegegoku** [DominoKit/domino-ui#965] [General]
Add your feedback and hints to help others who are migrating from 1.x.x to 2.x.x

**Comment by FrankHossfeld** [comment]
As a early adapter, I switch form Domino-UI v1 to v2 last summer. I am happy that I did the switch and it was really easy.  At that time a flex layout was created only with CSS and even that was no problem. Today, compared to last summer,  many bugs are fixed, the doc is much better, so, don't wait to upgrade ... :-)

**Comment by vegegoku** [comment]
### Migrating FlexLayout and FlexItem : 

Even thought the same syntax is preserved between the two versions to make migration easier, it is recommended to move to the new syntax when possible, users of the library can move to the new syntax gradually, the new syntax has some advantages over the old one, I will try to demonstrate that with few examples : 

- #### Shorter syntax : 
In version 1.x.x we can do this 

```java

FlexLayout.create()
    .setJustifyContent(FlexJustifyContent.SPACE_AROUND)
    .appendChild(FlexItem.create()
        .css(Styles.align_center)
        .appendChild(prevButton
            .addClickListener(evt -> {
              // go previous
            })
        )
    )
    .appendChild(FlexItem.create()
        .css(Styles.align_center)
        .appendChild(finishButton
            .addClickListener(evt -> {
              //do finish
            }))
    )
    .appendChild(FlexItem.create()
        .css(Styles.align_center)
        .appendChild(nextButton
            .addClickListener(evt -> {
              //go next
            }))
    )
```

in version 2.x.x 

```java
div().addCss(dui_flex, dui_justify_around)
        .appendChild(div().addCss(dui_center)
                .appendChild(prevButton.addClickListener(evt-> {
                    // go previous
                }))
        )
        .appendChild(div().addCss(dui_center)
                .appendChild(finishButton.addClickListener(evt-> {
                    // do finish
                }))
        )
        .appendChild(div().addCss(dui_center)
                .appendChild(nextButton.addClickListener(evt-> {
                    // go next
                }))
        );
```
#### FlexLayout not always a div: 
In version 1.x.x FlexLayout was always a div, but in v 2.x.x you can apply Flex styles to any type of elements.

#### Creating your custom layout and reuse them : 

With the introduction of the new css style interfaces in v2 it is now easier to create your custom flex layout and reuse them. for example if I want to make a vertical layout with specific gap between the items and align the items to the left, then reuse that in multiple places in my app

In version 1.x.x  you would make a custom component like this

```java
import elemental2.dom.HTMLDivElement;
import org.dominokit.domino.ui.shaded.grid.flex.FlexDirection;
import org.dominokit.domino.ui.shaded.grid.flex.FlexJustifyContent;
import org.dominokit.domino.ui.shaded.grid.flex.FlexLayout;
import org.dominokit.domino.ui.shaded.utils.BaseDominoElement;

public class VerticalLayout extends BaseDominoElement<HTMLDivElement, VerticalLayout> {

  private FlexLayout layout;

  public static VerticalLayout create() {
    return new VerticalLayout();
  }

  public VerticalLayout() {
    this.layout = FlexLayout.create()
        .setDirection(FlexDirection.TOP_TO_BOTTOM)
        .setGap("10px")
        .setJustifyContent(FlexJustifyContent.START);
    init(this);
  }

  @Override
  public HTMLDivElement element() {
    return this.layout.element();
  }
}
```
or you would introduce a static factory :

```java
  public static FlexLayout verticalLayout() {
    return FlexLayout.create()
        .setDirection(FlexDirection.TOP_TO_BOTTOM)
        .setGap("16px")
        .setJustifyContent(FlexJustifyContent.START);
  }
```

But in version 2.x.x you can use the composite css to achieve the same result way easier and way less code 
You just need to define a composite css class somewhere in your app like this

```java

public static CssClass vertical_layout = CompositeCssClass.of(dui_flex, dui_flex_col, dui_justify_start, dui_gap_4);
```

then just use it anywhere with any kind of element

```java
div().addCss(vertical_layout);
```

#### No need for FlexItem

In version 1.x.x any element or node that needs to be appended to a flex layout needed to be wrapped as a FlexItem, but this is no longer needed in version 2.x.x you can just use any element and you can use any element and append it directly to the flex element.

Example : 

In version 1.x.x 

```java
FlexLayout.create()
        .setDirection(FlexDirection.TOP_TO_BOTTOM)
        .setGap("10px")
        .setJustifyContent(FlexJustifyContent.START)
        .appendChild(FlexItem.create()
            .appendChild(div().textContent("first child"))
        )
        .appendChild(FlexItem.create()
            .appendChild(div().textContent("2nd child"))
        );
```

In version 2.x.x

```java
div().addCss(dui_flex, dui_flex_col, dui_justify_start, dui_gap_4)
                .appendChild(div().textContent("first child"))
                .appendChild(div().textContent("2nd child"));
```

These are some of the good qualities that are brought to the Flexlayout in version 2.x.x, some users might even find more advantages in the new syntax as it is extremely flexible and works well for building dynamic and complex layout, the only advantage of the old syntax over the new one could be the fact that the styles using enums and specific method for each might be easier, but this advantage starts to fade as you get used to the new syntax.

**Comment by vegegoku** [comment]
### Creating components with context styles : 

In version 1.x.x for some components when we needed to create an instance with a specific style context we needed to use one of the specified static factories for that component, for example : 

`Button.createPrimary("Submit")` or `Button.createDange("Cancel")`..etc those static method used to create the instance and attach a specific style  to the component.

In version 2.x.x we introduced instead context css classes that can be attached to any component instead, those styles will define a set of css variables that are used internally by the component to define how it looks, and instead of the static methods we can apply the style to the component by adding one of those context css classes, for example : 

`Button.create("Submit").addCss(dui_primary)` ,`Button.create("Cancel").addCss(dui_error)` 

The new way of applying those styles is a lot more flexible, because we can swap, toggle, or change those styles at well, plus they make it easier to theme the app or change them in runtime, and make it a lot easier to apply those to our own custom components or even any other components -knowing how to write our css to utilize them-.
